frontend/src/app/main/data/workspace/guides.cljs

```diff
@@ -0,0 +1,101 @@
+;; This Source Code Form is subject to the terms of the Mozilla Public
+;; License, v. 2.0. If a copy of the MPL was not distributed with this
+;; file, You can obtain one at http://mozilla.org/MPL/2.0/.
+;;
+;; Copyright (c) UXBOX Labs SL
+
+(ns app.main.data.workspace.guides
+  (:require
+   [app.common.geom.point :as gpt]
+   [app.common.geom.shapes :as gsh]
+   [app.common.spec :as us]
+   [app.common.types.page-options :as tpo]
+   [app.main.data.workspace.changes :as dwc]
+   [app.main.data.workspace.state-helpers :as wsh]
+   [beicon.core :as rx]
+   [cljs.spec.alpha :as s]
+   [potok.core :as ptk]))
+
+(defn make-update-guide [guide]
+  (fn [other]
+    (cond-> other
+      (= (:id other) (:id guide))
+      (merge guide))))
+
+(defn update-guides [guide]
+  (us/verify ::tpo/guide guide)
+  (ptk/reify ::update-guides
+    ptk/WatchEvent
+    (watch [it state _]
+      (let [page-id (:current-page-id state)
+            guides (-> state wsh/lookup-page-options (:guides {}))
+
+            new-guides (assoc guides (:id guide) guide)
+            
+            rch [{:type :set-option
+                  :page-id page-id
+                  :option :guides
+                  :value new-guides}]
+            uch [{:type :set-option
+                  :page-id page-id
+                  :option :guides
+                  :value guides}]]
+        (rx/of
+         (dwc/commit-changes
+          {:redo-changes rch
+           :undo-changes uch
+           :origin it}))))))
+
+(defn remove-guide [guide]
+  (us/verify ::tpo/guide guide)
+  (ptk/reify ::remove-guide
+    ptk/WatchEvent
+    (watch [it state _]
+      (let [page-id (:current-page-id state)
+            guides (-> state wsh/lookup-page-options (:guides {}))
+            new-guides (dissoc guides (:id guide))
+            
+            rch [{:type :set-option
+                  :page-id page-id
+                  :option :guides
+                  :value new-guides}]
+            uch [{:type :set-option
+                  :page-id page-id
+                  :option :guides
+                  :value guides}]]
+        (rx/of
+         (dwc/commit-changes
+          {:redo-changes rch
+           :undo-changes uch
+           :origin it}))))))
+
+(defn move-frame-guides
+  "Move guides that are inside a frame when that frame is moved"
+  [ids]
+  (us/verify (s/coll-of uuid?) ids)
+
+  (ptk/reify ::move-frame-guides
+    ptk/WatchEvent
+    (watch [_ state _]
+      (let [objects (wsh/lookup-page-objects state)
+            frame-ids? (->> ids (filter #(= :frame (get-in objects [% :type]))) (into #{}))
```

Comment:
you can retrieve the frames using `app.common.pages.helpers/select-frames`. Right now this function returns always a vector but maybe it can be defined like this to accept different collection type:

```clojure
(defn select-frames
  ([objects] (select-frames objects []))
  ([objects initial]
   (let [lookup #(get objects %)
         frame? #(= :frame (:type %))
         xform  (comp (map lookup)
                      (filter frame?))]
     (->> (:shapes (lookup uuid/zero))
          (into initial xform)))))
```

This is just a suggestion, I'm not sure if this change to the select-frames function is worth.

Reply:
I'm going to change it to this:
```clojure
            is-frame? (fn [id] (= :frame (get-in objects [id :type])))
            frame-ids? (into #{} (filter is-frame?) ids)
```